### PR TITLE
Localization: add base_map files for pyramid_map module

### PR DESCRIPTION
**Pull request by tthhee**

### CODE REVIEW

File: modules/localization/msf/local_pyramid_map/base_map/base_map.cc [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+  }
+  if (map_node_cache_lvl2_) {
+    delete map_node_cache_lvl2_;
+    map_node_cache_lvl2_ = nullptr;
+  }
+}
+
+void BaseMap::InitMapNodeCaches(int cacheL1_size, int cahceL2_size) {
+  if (map_node_cache_lvl1_ == nullptr) {
+    delete map_node_cache_lvl1_;
```
Comment (michael4338) [inline modules/localization/msf/local_pyramid_map/base_map/base_map.cc]: why delete a nullptr?  do you mean != ?

File: modules/localization/msf/local_pyramid_map/base_map/base_map.cc [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+    ++itr;
+  }
+
+  for (auto& future : load_futures_) {
+    if (future.valid()) {
+        future.get();
+    }
+  }
+
+  // check in cacheL2 again
```
Comment (michael4338) [inline modules/localization/msf/local_pyramid_map/base_map/base_map.cc]: consider moving duplicate code to a common function (not critical)

File: modules/localization/msf/local_pyramid_map/base_map/base_map.cc [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+    if (it->second != all_map_node_md5s_[i]) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool BaseMap::CheckMapStrictly() {
+  // TO DO
```
Comment (michael4338) [inline modules/localization/msf/local_pyramid_map/base_map/base_map.cc]: TODO(someone's name)

File: modules/localization/msf/local_pyramid_map/base_map/base_map.cc (hunk 1 above)
Comment (tthhee) [reply modules/localization/msf/local_pyramid_map/base_map/base_map.cc]: thanks, it's a mistake. already fixed

File: modules/localization/msf/local_pyramid_map/base_map/base_map.cc (hunk 2 above)
Comment (tthhee) [reply modules/localization/msf/local_pyramid_map/base_map/base_map.cc]: fixed

File: modules/localization/msf/local_pyramid_map/base_map/base_map.cc (hunk 3 above)
Comment (tthhee) [reply modules/localization/msf/local_pyramid_map/base_map/base_map.cc]: fixed

File: modules/localization/msf/local_pyramid_map/base_map/base_map.cc [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+    delete map_node_cache_lvl1_;
+    map_node_cache_lvl1_ == nullptr;
+  }
+
+  if (map_node_cache_lvl2_ != nullptr) {
+    delete map_node_cache_lvl2_;
+    map_node_cache_lvl2_ == nullptr;
+  }
+
+  map_node_cache_lvl1_ =
```
Comment (jinghaomiao) [inline modules/localization/msf/local_pyramid_map/base_map/base_map.cc]: Consider using shared pointer to avoid manual deletion?

Comment (tthhee) [reply modules/localization/msf/local_pyramid_map/base_map/base_map.cc]: fixed

File: modules/localization/msf/local_pyramid_map/base_map/base_map_pool.h [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  /**@brief Get the size of pool. */
+  unsigned int GetPoolSize() { return pool_size_; }
+
+ private:
+  /**@brief The task function of the thread pool for release node.
+   * @param <map_node> The released MapNode object.
+   * */
+  void FreeMapNodeTask(BaseMapNode* map_node);
+  /**@brief new a map node. */
+  virtual BaseMapNode* AllocNewMapNode() = 0;
```
[the remark this answers is not in the document]
Comment (tthhee) [reply modules/localization/msf/local_pyramid_map/base_map/base_map_pool.h]: we need to control the release of pointer in base_map_pool

File: modules/localization/msf/local_pyramid_map/base_map/base_map_pool.cc [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    }
+  }
+  boost::unique_lock<boost::mutex> lock(mutex_);
+  if (free_list_.empty()) {
+    if (is_fixed_size_) {
+      return NULL;
+    }
+    BaseMapNode* node = AllocNewMapNode();
+    InitNewMapNode(node);
+    pool_size_++;
```
[the remark this answers is not in the document]
Comment (tthhee) [reply modules/localization/msf/local_pyramid_map/base_map/base_map_pool.cc]: fixed

File: modules/localization/msf/local_pyramid_map/base_map/base_map_pool.cc [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+void BaseMapNodePool::InitNewMapNode(BaseMapNode* node) {
+  node->Init(map_config_);
+  return;
+}
+
+void BaseMapNodePool::FinalizeMapNode(BaseMapNode* node) { node->Finalize(); }
+
+void BaseMapNodePool::DellocMapNode(BaseMapNode* node) { delete node; }
+
+void BaseMapNodePool::ResetMapNode(BaseMapNode* node) { node->ResetMapNode(); }
```
[the remark this answers is not in the document]
Comment (tthhee) [reply modules/localization/msf/local_pyramid_map/base_map/base_map_pool.cc]: fixed

**Pull request merged**